packages/contracts/test/integration/commitReveal.spec.ts

```diff
@@ -0,0 +1,232 @@
+import * as ethers from "ethers";
+
+import * as Utils from "@counterfactual/test-utils";
+import Multisig from "../helpers/multisig";
+
+const CommitRevealApp = artifacts.require("CommitRevealApp");
+
+const web3 = (global as any).web3;
+const { provider, unlockedAccount } = Utils.setupTestEnv(web3);
+
+const [A, B] = [
+  // 0xaeF082d339D227646DB914f0cA9fF02c8544F30b
+  new ethers.Wallet(
+    "0x3570f77380e22f8dc2274d8fd33e7830cc2d29cf76804e8c21f4f7a6cc571d27",
+    provider
+  ),
+  // 0xb37e49bFC97A948617bF3B63BC6942BB15285715
+  new ethers.Wallet(
+    "0x4ccac8b1e81fb18a98bbaf29b9bfe307885561f71b76bd4680d7aec9d0ddfcfd",
+    provider
+  )
+];
+
+const computeStateHash = (
+  appStateHash: string,
+  nonce: number,
+  timeout: number
+) =>
+  ethers.utils.solidityKeccak256(
+    ["bytes1", "address[]", "uint256", "uint256", "bytes32"],
+    ["0x19", [A.address, B.address], nonce, timeout, appStateHash]
+  );
+
+const encode = (encoding: string, state: any) =>
+  ethers.utils.defaultAbiCoder.encode([encoding], [state]);
+
+function keccak256Struct(encoding: string, struct: any) {
+  const bytes = ethers.utils.defaultAbiCoder.encode([encoding], [struct]);
+  return ethers.utils.solidityKeccak256(["bytes"], [bytes]);
+}
+
+enum AssetType {
+  ETH,
+  ERC20
+}
+
+enum Stage {
+  SETTING_MAX,
+  CHOOSING,
+  GUESSING,
+  REVEALING,
+  DONE
+}
+
+enum Player {
+  CHOOSING = 0,
+  GUESSING = 1
+}
+
+describe("CommitRevealApp", async () => {
+  const StateChannel = artifacts.require("StateChannel");
+  const ConditionalTransfer = artifacts.require("ConditionalTransfer");
+  const StaticCall = artifacts.require("StaticCall");
+  const NonceRegistry = artifacts.require("NonceRegistry");
+  const Registry = artifacts.require("Registry");
+  const Signatures = artifacts.require("Signatures");
+  const Transfer = artifacts.require("Transfer");
+
+  let app;
+
+  const appEncoding =
+    "tuple(address addr, bytes4 reducer, bytes4 resolver, bytes4 turnTaker, bytes4 isStateFinal)";
+
+  const appStateEncoding =
+    "tuple(address[2] playerAddrs, uint256 stage, uint256 maximum, " +
+    "uint256 guessedNumber, bytes32 commitHash, uint256 winner)";
+
+  const termsEncoding = "tuple(uint8 assetType, uint256 limit, address token)";
+
+  beforeEach(async () => {
+    CommitRevealApp.link("StaticCall", StaticCall.address);
+
+    StateChannel.link("Signatures", Signatures.address);
+    StateChannel.link("StaticCall", StaticCall.address);
+    StateChannel.link("Transfer", Transfer.address);
+  });
+
+  it("should complete a full lifecycle", async function() {
```

Comment:
This test is ~142 lines long. 

We can break it up and make it more readable and understandable.

Easy hack is to just take  all the numbered comments and make a function out of it where the name of the func is the comment. Then the code is self documenting.

Reply:
The test is definitely a bit too long. I've written some test tooling that's going to reduce the # of lines here.

I don't like the extract-to-function approach, it removes explicitness.

Reply:
Accepting since these aren't critical issues to block on if we disagree, but they are interesting nonetheless,

Can you explain what you mean about explicitness?

The extract-to-function approach is only a heuristic. Let me give some context into how I think about this.

The point is that, when I'm reading tests,  I'm often trying to understand what a particular function or class is doing. I want a high level overview and so I want to operate at one abstraction level, and only step down abstraction levels as needed. I don't want to read the implementation details of how a contract is deployed, or what `setState` means.

I'm often just trying to figure out, what tests are doing? What are they not doing? How do I use this class? 

These are useful questions from the perspective of a new person reading the code. However, when we have monolithic tests like this, it makes answering these questions more difficult.

Take for example, the lifecycle test in the machine:
```
it("should exercise the lifecycle of a state channel", async () => {
  //given
  let [walletA, walletB] = createWallet();

  //when
  await setup(walletA, walletB);
  await makeDeposits(walletA, walletB);
  await playTtt(walletA, walletB);

  // then
  validateSystem(walletA, walletB);
  await gotoChain();
  await validateBlockchain();
}
```

This is a huge test with lots of code, but when reading the test function, you can immediately understand the entirety of what it does. If you care about how TTT is played, you can just go into that function if you want.

Generally, I lean more towards the direction of uncle bob's "CleanCode" style. This makes for a lot of helper functions and classes (via SRP), but makes code read more like English.

Reply:
I see your point. 

However, to me it feels like there's a context-switching cost to jumping between functions. I find myself constantly having to remember which stage of the test I'm in and which data is currently being operated on. 

By "explicitness" I mean transparency of the code that's executed. More explicit tests might have a steeper learning curve at first, but I find they are better for understanding the logic being tested more deeply. 

That said, I'm not certain. I'll try understandable-at-a-glance-style tests in a later PR, and compare.

Reply:
Fwiw, my philosophy here isn't just for test. It applies to all code.

Reply:
Yeah, and in general i agree with you. But there's a time and a place for both philosophies IMO